### ISSUE TITLE
Clarification: can a service be registered/unregistered more than once?

### ISSUE THREAD

**Issue by angusholder** [Zhuinden/simple-stack#264]
I'm trying to write something equivalent to `val ViewModel.viewModelScope: CoroutineScope`:

```kotlin
class FooViewModel : ScopedServices.Registered {
    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    override fun onServiceRegistered() {
    }

    override fun onServiceUnregistered() {
        serviceScope.cancel()
    }
}
```

Am I correct in thinking that once unregistered, the service is destroyed and that instance won't ever be reused? So `onServiceRegistered` and `onServiceUnregistered` only ever get called once each?

Another question; with Jetpack ViewModel, I often have a `loadInitialData()` method that I call from `Fragment.onViewCreated`, because I'd rather not kick off database calls/network calls from the constructor, it feels hacky and is not nice for testing. `onServiceRegistered` looks ideal for kicking off initial data loading, would you recommend that?

**Comment by Zhuinden** [comment]
> Am I correct in thinking that once unregistered, the service is destroyed and that instance won't ever be reused?

Back in **Simple-Stack 1.x** there was a "bug" (design oversight) caused by how re-registering the same instance would call `onEnterScope()` to run multiple times, **2,x** exists specifically because this section was heavily reworked to ensure that 1 object is tracked only once per registration.

So even if the same instance is added in multiple scopes, it would only have `onServiceRegistered` called when `0 -> 1`, and `onServiceUnregistered()` only called when `1 -> 0`, but not later.

Therefore, unless you cache these poor objects in some singleton and "re-add the same instance" again, then these callbacks are ensured to be only called once.

TL;DR:

> Am I correct in thinking that once unregistered, the service is destroyed and that instance won't ever be reused? So onServiceRegistered and onServiceUnregistered only ever get called once each?

Yes

> Another question; with Jetpack ViewModel, I often have a loadInitialData() method that I call from Fragment.onViewCreated, because I'd rather not kick off database calls/network calls from the constructor, it feels hacky and is not nice for testing. onServiceRegistered looks ideal for kicking off initial data loading, would you recommend that?

Yes

😉

**Comment by angusholder** [comment]
Okay, that's helpful thank you. Want me to make a PR to add that info onto `onServiceRegistered`/`onServiceUnregistered`?

**Comment by Zhuinden** [comment]
Hmm, there is already a comment like so:

```java
    /**
     * When a service implements {@link Registered}, then it will receive a callback when the service has been registered to at least one scope.
     */
    public static interface Registered {
        /**
         * Called when the service has been registered to any scopes, and this is the first scope it was registered to.
         */
        void onServiceRegistered();

        /**
         * Called when the service is no longer registered in any scopes.
         */
        void onServiceUnregistered();
    }
```

Although if it's not actually lcear, then it can be useful. It also doesn't really mention how `onServiceRegisterd` runs after `fromBundle`.

**Comment by Zhuinden** [comment]
I will consider this solved.

**Comment by angusholder** [comment]
Sorry I didn't get back to you, yeah I think it's fine
